pkg/networkservice/common/crossnse/server.go

```diff
@@ -0,0 +1,172 @@
+// their unix file socket as the clienturl.ClientURL(ctx) used to connect to them.
+package crossnse
+
+import (
+	"context"
+	"net/url"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/trace"
+	"github.com/networkservicemesh/sdk/pkg/registry/common/crossnse"
+
+	"github.com/golang/protobuf/ptypes/empty"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type crossConnectNSEServer struct {
+	// Map of names -> *registry.NetworkServiceEndpoint for local bypass to file crossNSEs
+	crossNSEs sync.Map
+
+	activeConnection sync.Map
+
+	nsmgrName string
+	nsmgrInit sync.Once
+}
+
+type connectionInfo struct {
+	endpointURL *url.URL
+	crossNSEURL *url.URL
+	closingNSE  bool
+}
+
+// NewServer - creates a NetworkServiceServer that tracks locally registered CrossConnect Endpoints and on first Request forward to cross conenct nse
+//				one by one and if request came back from cross nse, it will connect to a proper next client endpoint.
+//             - server - *registry.NetworkServiceRegistryServer.  Since registry.NetworkServiceRegistryServer is an interface
+//                        (and thus a pointer) *registry.NetworkServiceRegistryServer is a double pointer.  Meaning it
+//                        points to a place that points to a place that implements registry.NetworkServiceRegistryServer
+//                        This is done so that we can return a registry.NetworkServiceRegistryServer chain element
+//                        while maintaining the NewServer pattern for use like anything else in a chain.
+//                        The value in *server must be included in the registry.NetworkServiceRegistryServer listening
+//                        so it can capture the registrations.
+func NewServer(registryServer *registry.NetworkServiceEndpointRegistryServer) networkservice.NetworkServiceServer {
+	rv := &crossConnectNSEServer{}
+	*registryServer = crossnse.NewNetworkServiceRegistryServer(rv)
+	return rv
+}
+
+func (l *crossConnectNSEServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (result *networkservice.Connection, err error) {
+	// Check if there is no active connection, we need to replace endpoint url with forwarder url
+	conn := request.GetConnection()
+	ind := conn.GetPath().GetIndex() // It is designed to be used inside Endpoint, so current index is Endpoint already
+	connID := conn.GetId()
+
+	if len(conn.GetPath().GetPathSegments()) == 0 || ind <= 0 {
+		return nil, errors.Errorf("path segment doesn't have a client or cross connect nse identity")
+	}
+	// Remember current segment Name if not yet.
+	l.nsmgrInit.Do(func() {
+		l.nsmgrName = conn.GetPath().GetPathSegments()[ind].Name
```

Comment:
I presume this is intended to happen after updatepath and so you are piggy backing on update path knowing the name?

Reply:
Fixed, passed name to chain element, to simplify.